ledger/simulation/resources.go

```diff
@@ -0,0 +1,391 @@
+	}
+	return a.localTxnResources[txnIndex].hasAsset(aid)
+}
+
+func (a *GroupResourceAssignment) addAsset(aid basics.AssetIndex, globalSharing bool, txnIndex int) bool {
+	if globalSharing {
+		return a.Resources.addAsset(aid)
+	}
+	return a.localTxnResources[txnIndex].addAsset(aid)
+}
+
+func (a *GroupResourceAssignment) hasApp(aid basics.AppIndex, globalSharing bool, txnIndex int) bool {
+	if globalSharing {
+		return a.Resources.hasApp(aid)
+	}
+	return a.localTxnResources[txnIndex].hasApp(aid)
+}
+
+func (a *GroupResourceAssignment) addApp(aid basics.AppIndex, ep *logic.EvalParams, programVersion uint64, globalSharing bool, txnIndex int) bool {
+	if globalSharing {
+		return a.Resources.addApp(aid, ep, programVersion)
+	}
+	return a.localTxnResources[txnIndex].addApp(aid, ep, programVersion)
+}
+
+func (a *GroupResourceAssignment) hasBox(app basics.AppIndex, name string) bool {
+	// all boxes are global, never consult PerTxnResources
+	return a.Resources.hasBox(app, name)
+}
+
+func (a *GroupResourceAssignment) addBox(app basics.AppIndex, name string) bool {
+	// all boxes are global, never consult PerTxnResources
+	return a.Resources.addBox(app, name)
+}
+
+func (a *GroupResourceAssignment) hasHolding(addr basics.Address, aid basics.AssetIndex) bool {
+	// nil map lookup is ok
+	_, ok := a.AssetHoldings[ledgercore.AccountAsset{Address: addr, Asset: aid}]
+	return ok
+}
+
+func (a *GroupResourceAssignment) addHolding(addr basics.Address, aid basics.AssetIndex) bool {
+	// TODO: limit cross-product usage
+	if a.AssetHoldings == nil {
+		a.AssetHoldings = make(map[ledgercore.AccountAsset]struct{})
+	}
+	a.AssetHoldings[ledgercore.AccountAsset{Address: addr, Asset: aid}] = struct{}{}
+	return true
+}
+
+func (a *GroupResourceAssignment) hasLocal(addr basics.Address, aid basics.AppIndex, ep *logic.EvalParams) bool {
+	if ep.GetApplicationAddress(aid) == addr {
+		// The app local of an app and its own account is always available, so don't bother recording it.
+		return true
+	}
+	// nil map lookup is ok
+	_, ok := a.AppLocals[ledgercore.AccountApp{Address: addr, App: aid}]
+	return ok
+}
+
+func (a *GroupResourceAssignment) addLocal(addr basics.Address, aid basics.AppIndex) bool {
+	// TODO: limit cross-product usage
+	if a.AppLocals == nil {
+		a.AppLocals = make(map[ledgercore.AccountApp]struct{})
+	}
+	a.AppLocals[ledgercore.AccountApp{Address: addr, App: aid}] = struct{}{}
+	return true
+}
+
+func (a *GroupResourceAssignment) boxIOBudget(bytesPerBoxRef uint64) uint64 {
+	return uint64(a.startingBoxes+a.Resources.maxPossibleBoxes()) * bytesPerBoxRef
+}
+
+type resourcePolicy struct {
+	assignment                  GroupResourceAssignment
+	ep                          *logic.EvalParams
+	initialBoxSurplusReadBudget *uint64
+	unnamedBoxUsedReadBudget    uint64
+
+	txnRootIndex   int
+	programVersion uint64
+	globalSharing  bool
+}
+
+func newResourcePolicy(ep *logic.EvalParams, groupResult *TxnGroupResult) *resourcePolicy {
```

Comment:
so I assume this is where the resource recording starts: before we step into txn group, we create the `resourcePolicy` instance and bury the corresponding parts into txn group result.

As we execute the txn group, we append/check resource in `assignment` field. By the end of execution, we collect all the resources in right places.

looks great.